test/services/RPCSelector.test.ts

```diff
@@ -19,25 +21,111 @@ describe('RPCSelector', () => {
+        it('returns the preferred URL', async () => {
+          const mockedProvider = sinon.stub(JsonRpcProvider.prototype, 'getBlock').resolves(stubValue);
+
+          const urls = [
+            'https://data-seed-prebsc-1-s2.binance.org:8545/',
+            'https://data-seed-prebsc-2-s3.binance.org:8545/',
+          ];
+          const rpcSelector = new RPCSelector(urls);
+          expect(await rpcSelector.apply()).to.eq(urls[0]);
+
+          mockedProvider.restore();
+        });
       });
-    });
 
-    describe('when the preferred URL is not up to date', () => {
-      it('returns the RPC with highest block number', async () => {
-        const urls = ['https://data-seed-prebsc-2-s1.binance.org:8545/','https://data-seed-prebsc-1-s1.binance.org:8545/'];
-        const rpcSelector = new RPCSelector(urls);
-        expect(await rpcSelector.apply()).to.eq(urls[1]);
+      describe('when the preferred URL is not up to date', () => {
+        const stubValue = {
+          timestamp: Math.floor(Date.now() / 1000) - 120,
+        } as Block;
+
+        it('returns the RPC with highest block number', async () => {
+          const mockedGetBlock = sinon.stub(JsonRpcProvider.prototype, 'getBlock').resolves(stubValue);
+          const mockedGetBlockNumber = sinon.stub(JsonRpcProvider.prototype, 'getBlockNumber').resolves(100_000);
+
+          const urls = [
+            'https://data-seed-prebsc-1-s1.binance.org:8545/',
+            'https://data-seed-prebsc-1-s2.binance.org:8545/',
+          ];
+          const rpcSelector = new RPCSelector(urls, 100);
+          expect(await rpcSelector.apply()).to.eq(urls[1]);
+          expect(mockedGetBlock.called);
+
+          mockedGetBlock.restore();
+          mockedGetBlockNumber.restore();
+        });
+      });
+
+      describe('when RPC exceeds request time threshold', () => {
+        const fakeGetBlock = (): Promise<Block> => {
+          return new Promise((resolve) => {
+            setTimeout(resolve, 5000, { timestamp: Math.floor(Date.now() / 1000) });
+          });
+        };
+
+        const fakeGetBlockNumber = (): Promise<number> => {
+          return new Promise((resolve) => {
+            setTimeout(resolve, 5000, 100_000);
+          });
+        };
+
+        it('timeouts and select the first of the non-preferred RPCs', async () => {
+          const mockedGetBlock = sinon.stub(JsonRpcProvider.prototype, 'getBlock').callsFake(fakeGetBlock);
+          const mockedGetBlockNumber = sinon
+            .stub(JsonRpcProvider.prototype, 'getBlockNumber')
+            .callsFake(fakeGetBlockNumber);
+
+          const urls = [
+            'https://data-seed-prebsc-1-s1.binance.org:8545/',
+            'https://data-seed-prebsc-1-s2.binance.org:8545/',
+          ];
+
+          const rpcSelector = new RPCSelector(urls, 500);
+          expect(await rpcSelector.apply()).to.eq(urls[1]);
+
+          mockedGetBlock.restore();
+          mockedGetBlockNumber.restore();
+        });
```

Comment:
which of the two timed out?

Reply:
It timeouts both. The reduce will take care of selecting the first one of the non-preferred URLs.